docs/config/node_config.md
# 节点配置


## Worker访问端口

### WORKER_PORT 【开放】

通过该端口，外部可以访问到worker服务。（注意无法访问控制台）
```
WORKER_PORT=8080
```

## 控制台端口

通过这些配置，可以访问到控制台。

### MASTER_ENDPOINT

```
MASTER_ENDPOINT=http://127.0.0.1:8888  
```
该地址为master节点的地址，子节点通过该地址与master节点进行通信

### API_PORT【开放】

```
API_PORT=8888
```
该端口为开放的控制台的端口，在该端口上显示控制台网页。

以上两个端口应当保持一致（除非被容器映射成不同的端口）。

## Worker网络转发端口

### TUNNEL_HOST

```
TUNNEL_HOST=127.0.0.1
```
该地址为master节点的地址，子节点通过该地址与master节点进行内部通信。

主节点保持为127.0.0.1，子节点则为master节点的ip。

### TUNNEL_ENTRY_PORT

```
TUNNEL_ENTRY_PORT=10080
```
主节点的内部端口，如非必要请勿配置，子节点无法访问，主节点用该端口向子节点发送请求。

### TUNNEL_API_PORT【开放】

```
TUNNEL_API_PORT=18080
```
主节点应当公开给子节点的端口，子节点通过该端口配置每个worker的网络。


## Worker配置

### WORKER_URL_SUFFIX

```
WORKER_URL_SUFFIX=.example.com
```
配置域名后缀，如`.example.com`，则worker的域名将为`WORKERNAME+WORKER_URL_SUFFIX`，如`worker.example.com`。

### SCHEME

```
SCHEME=http
```
http/https，从网页访问的协议，可能影响跨域。


### NODE_NAME

```
NODE_NAME=default
```
节点名称，主节点默认为`default`，子节点需要配置的与众不同。

### AGENT_SECRET

```
AGENT_SECRET=123123
```
子节点注册master时需要的密钥，不同节点的密钥应当保持一致。

### RUN_MODE

```
RUN_MODE=master
```
主节点为`master`，子节点为`agent`。

### WORKER_HOST_MODE

```
WORKER_HOST_MODE=host
```
host/path，host模式下，worker的url为`workername.vvorker.domain`，path模式下，worker的url为`vvorker.domain/workername`。

### WORKER_HOST_PATH

```
WORKER_HOST_PATH=something
```
在path模式下，worker的url为`vvorker.domain/WORKER_HOST_PATH/workername`。在某些反向代理的情况下可能有用。

### ADMIN_API_PROXY

```
ADMIN_API_PROXY=false
```
允许admin页面代理api请求，这可能会导致路径冲突，并且WORKER_HOST_MODE必须为path。这将允许一个端口同时提供admin页面和worker服务。不推荐使用。

### DB_TYPE

```
DB_TYPE=sqlite
```
数据库类型，sqlite/postgres/mysql。
这将选择平台使用的数据库类型，注意需要配置各个数据库的连接信息。

### DB_PATH【重要】
当使用sqlite时

```
DB_PATH=/app/data/db.sqlite
```
sqlite数据库的路径。需要针对操作系统配置合适的路径。

### DB_NAME

```
DB_NAME=vvorker
```
当使用postgres/mysql时，可以指定系统数据库的名称。对于多个节点使用同一个数据库服务时较为有用。

### WORKERD_DIR【重要】

```
WORKERD_DIR=/app/data
```
workerd的目录，用于存储各个worker的运行时文件。无需持久化，仅作为缓存使用。需要针对操作系统配置合适的路径。

### WORKERD_BIN_PATH【重要】

```
WORKERD_BIN_PATH=/bin/workerd
```
workerd的二进制文件路径。需要对应位置有workerd的二进制文件，如`.exe`，注意是二进制的，而不是脚本文件。使用npm安装workerd时需要格外注意。


### API_WEB_BASE_URL

```
API_WEB_BASE_URL=http://127.0.0.1:8080
```
用于指定控制台中快速跳转服务的地址前缀。在path模式下使用。

### COOKIE_NAME

```
COOKIE_NAME=vv-authorization
```
用于指定admin的cookie的名称。

### COOKIE_AGE

```
COOKIE_AGE=86400
```
cookie的过期时间，单位为秒。

### COOKIE_DOMAIN

```
COOKIE_DOMAIN=vvorker.local
```
cookie的域名。

### ENABLE_REGISTER

```
ENABLE_REGISTER=false
```
是否允许注册。

### JWT_SECRET

```
JWT_SECRET=123123
```
jwt的密钥。

## KV 配置

### KV_PROVIDER

```
KV_PROVIDER=redis
```
kv的提供者，redis/nutsdb。选择redis通常是理想的情况，但还需要安装redis；nutsdb则为本地存储，不需要进行安装。

### SERVER_REDIS_HOST

```
SERVER_REDIS_HOST=127.0.0.1
```
redis的地址。

### SERVER_REDIS_PORT

```
SERVER_REDIS_PORT=6379
```
redis的端口。

## OSS

### SERVER_MINIO_HOST

```
SERVER_MINIO_HOST=127.0.0.1
```
minio的地址。

### SERVER_MINIO_PORT

```
SERVER_MINIO_PORT=9000
```
minio的端口。

### SERVER_MINIO_REGION

```
MINIO_REGION=us-east-1
```
minio的区域。

### SERVER_MINIO_ACCESS

```
MINIO_ACCESS=minioadmin
```
minio的访问密钥。

### SERVER_MINIO_SECRET

```
MINIO_SECRET=minioadmin
```
minio的访问密钥。

### SERVER_MINIO_USE_SSL

```
MINIO_USE_SSL=false
```
是否使用https。

### MINIO_SINGLE_BUCKET_MODE

```
MINIO_SINGLE_BUCKET_MODE=false
```
是否使用单个bucket，所有应用都使用同一个bucket下的不同文件夹，注意，这将不进行权限管控。

### MINIO_SINGLE_BUCKET_NAME

```
MINIO_SINGLE_BUCKET_NAME=vvorker
```
如果使用单个bucket，bucket名称。

## PostgreSQL

### SERVER_POSTGRE_HOST

```
SERVER_POSTGRE_HOST=127.0.0.1
```
postgres的地址。

### SERVER_POSTGRE_PORT

```
SERVER_POSTGRE_PORT=5432
```
postgres的端口。

### SERVER_POSTGRE_USER

```
SERVER_POSTGRE_USER=postgres
```
postgres的用户。

### SERVER_POSTGRE_PASSWORD

```
SERVER_POSTGRE_PASSWORD=postgres
```
postgres的密码。

## MySQL

### SERVER_MYSQL_HOST

```
SERVER_MYSQL_HOST=127.0.0.1
```
mysql的地址。

### SERVER_MYSQL_PORT

```
SERVER_MYSQL_PORT=3306
```
mysql的端口。

### SERVER_MYSQL_USER

```
SERVER_MYSQL_USER=root
```
mysql的用户。

### SERVER_MYSQL_PASSWORD

```
SERVER_MYSQL_PASSWORD=root
```
mysql的密码。

### SERVER_MYSQL_ONE_DB_NAME

```
SERVER_MYSQL_ONE_DB_NAME=vvorker
```
当不为空时，所有mysql资源都将在同一个库中，并且不进行权限控制。

## 本地内部端口

注意这些端口不要重复即可，通常一台机器上不使用容器部署多个节点时，需要配置不同的端口。

### CLIENT_POSTGRE_PORT

```
CLIENT_POSTGRE_PORT=15432
```
postgres的端口。

### CLIENT_MYSQL_PORT

```
CLIENT_MYSQL_PORT=15433
```
mysql的端口。

### CLIENT_REDIS_PORT

```
CLIENT_REDIS_PORT=16379
```
redis的端口。

### LOCAL_TMP_POSTGRE_PORT

```
LOCAL_TMP_POSTGRE_PORT=13420
```
tmp postgres的端口。

### LOCAL_TMP_REDIS_PORT

```
LOCAL_TMP_REDIS_PORT=13421
```
tmp redis的端口。

### LOCAL_TMP_MYSQL_PORT

```
LOCAL_TMP_MYSQL_PORT=13422
```
tmp mysql的端口。

## SSO

### SSO_AUTH_URL

```
SSO_AUTH_URL=http://127.0.0.1:8888/sso/auth
```
sso认证地址。

### SSO_REDIRECT_URL

```
SSO_REDIRECT_URL=http://127.0.0.1:8888/sso/redirect
```
登录页。

### SSO_BASE_URL

```
SSO_BASE_URL=http://127.0.0.1:8888
```
sso基础地址。`xxxx.com/WORKER_HOST_PATH/SSO_BASE_URL/WORKER_NAME`
在服务器不配置WORKER_HOST_PATH时，但前端需要BASE_URL时可能有用。

### SSO_COOKIE_NAME

```
SSO_COOKIE_NAME=vv-sso
```
sso cookie名称。


conf/env.go
package conf

import (
	"fmt"
	"os"
	"vvorker/utils/secret"

	"github.com/golang-jwt/jwt/v5"
	"github.com/ilyakaznacheev/cleanenv"
	"github.com/joho/godotenv"
	"github.com/sirupsen/logrus"
)

type AppConfig struct {
	MasterEndpoint string `env:"MASTER_ENDPOINT" env-default:"http://127.0.0.1:8888"` // needed for agent，agent需要通过该url来注册节点
	WorkerPort     int    `env:"WORKER_PORT" env-default:"8080"`                      // 【主节点公开】提供worker服务，如 xxx.example.com:8080
	APIPort        int    `env:"API_PORT" env-default:"8888"`                         // 【主节点公开】提供控制台服务和控制api服务

	TunnelHost      string `env:"TUNNEL_HOST" env-default:"127.0.0.1"`   // for master usually 127.0.0.1, for agent usually master public ip
	TunnelEntryPort int    `env:"TUNNEL_ENTRY_PORT" env-default:"10080"` // 【主节点内部】提供http服务，主节点用这个端口向其他节点发送请求。在主节点提供服务，子节点无法向其发送请求。
	TunnelAPIPort   int    `env:"TUNNEL_API_PORT" env-default:"18080"`   // 【主节点公开】TUNNEL_ENTRY_PORT的frp端口，子节点通过这个端口配置每个worker的转发

	InternalRPCPort int `env:"INTERNAL_RPC_PORT" env-default:"19080"` // 【内部】提供worker之间的rpc服务，未实现

	WorkerURLSuffix string `env:"WORKER_URL_SUFFIX" env-default:".vvorker.local"` // master required, e.g. .example.com. for worker show and route
	Scheme          string `env:"SCHEME" env-default:"http"`                      // http, https. for public frontend show
	NodeName        string `env:"NODE_NAME" env-default:"default"`
	AgentSecret     string `env:"AGENT_SECRET"` //	required, e.g. 123123123

	DBPath         string `env:"DB_PATH" env-default:"/app/data/db.sqlite"`
	WorkerdDir     string `env:"WORKERD_DIR" env-default:"/app/data"`
	DBType         string `env:"DB_TYPE" env-default:"sqlite"`
	DBName         string `env:"DB_NAME" env-default:"vvorker"`
	WorkerLimit    int    `env:"WORKER_LIMIT" env-default:"10000"`
	WorkerdBinPath string `env:"WORKERD_BIN_PATH" env-default:"/bin/workerd"`

	EncryptionKey string `env:"ENCRYPTION_KEY" env-default:""`

	APIWebBaseURL  string `env:"API_WEB_BASE_URL"`
	ListenAddr     string `env:"LISTEN_ADDR" env-default:"0.0.0.0"`
	CookieName     string `env:"COOKIE_NAME" env-default:"vv-authorization"`
	CookieAge      int    `env:"COOKIE_AGE" env-default:"86400"`            // second 86400 = 1 day
	CookieDomain   string `env:"COOKIE_DOMAIN" env-default:"vvorker.local"` // required, e.g. example.com
	EnableRegister bool   `env:"ENABLE_REGISTER" env-default:"false"`
	RunMode        string `env:"RUN_MODE" env-default:"master"` // master, agent

	LitefsPrimaryPort int    `env:"LITEFS_PRIMARY_PORT" env-default:"20202"`
	LitefsBinPath     string `env:"LITEFS_BIN_PATH" env-default:"/usr/local/bin/litefs"`
	LitefsDirPath     string `env:"LITEFS_DIR_PATH" env-default:"/app"`
	LitefsEnabled     bool   `env:"LITEFS_ENABLED" env-default:"false"`
	EnableAutoSync    bool   `env:"ENABLE_AUTO_SYNC" env-default:"false"`
	TunnelUsername    string
	TunnelPassword    string
	TunnelToken       string
	NodeID            string

	KVProvider string `env:"KV_PROVIDER" env-default:"redis"` // redis nutsdb

	WorkerHostMode string `env:"WORKER_HOST_MODE" env-default:"host"` // host path  // host 模式需要使用域名进行访问，path则url的第一段为服务名（不包含域名后缀
	WorkerHostPath string `env:"WORKER_HOST_PATH" env-default:""`     // host 模式需要使用域名进行访问，path则url的第一段为服务名（不包含域名后缀，如example.com/xxxx/admin
	AdminAPIProxy  bool   `env:"ADMIN_API_PROXY" env-default:"false"` // 允许admin页面代理api请求，这可能会导致路径冲突，并且WORKER_HOST_MODE必须为path

	ServerRedisHost string `env:"SERVER_REDIS_HOST" env-default:"localhost"`
	ServerRedisPort int    `env:"SERVER_REDIS_PORT" env-default:"6379"`

	ServerMinioHost   string `env:"SERVER_MINIO_HOST" env-default:"localhost"` // localhost时为本地
	ServerMinioPort   int    `env:"SERVER_MINIO_PORT" env-default:"9000"`      // 本地时为9000，远程时为443
	ServerMinioRegion string `env:"SERVER_MINIO_REGION" env-default:"us-east-1"`
	ServerMinioUseSSL bool   `env:"SERVER_MINIO_USE_SSL" env-default:"false"`
	ServerMinioAccess string `env:"SERVER_MINIO_ACCESS" env-default:"minioadmin"`
	ServerMinioSecret string `env:"SERVER_MINIO_SECRET" env-default:"minioadmin"`

	MinioSingleBucketMode bool   `env:"MINIO_SINGLE_BUCKET_MODE" env-default:"false"`   // 是否使用单个bucket，所有应用都使用同一个bucket下的不同文件夹，注意，这将不进行权限管控
	MinioSingleBucketName string `env:"MINIO_SINGLE_BUCKET_NAME" env-default:"vvorker"` // 如果使用单个bucket，bucket名称

	ServerPostgreHost     string `env:"SERVER_POSTGRE_HOST" env-default:"localhost"`
	ServerPostgrePort     int    `env:"SERVER_POSTGRE_PORT" env-default:"5432"`
	ServerPostgrePassword string `env:"SERVER_POSTGRE_PASSWORD" env-default:"postgres"`
	ServerPostgreUser     string `env:"SERVER_POSTGRE_USER" env-default:"postgres"`

	ServerMySQLHost      string `env:"SERVER_MYSQL_HOST" env-default:"localhost"`
	ServerMySQLPort      int    `env:"SERVER_MYSQL_PORT" env-default:"3306"`
	ServerMySQLPassword  string `env:"SERVER_MYSQL_PASSWORD" env-default:"root123"`
	ServerMySQLUser      string `env:"SERVER_MYSQL_USER" env-default:"root"`
	ServerMySQLOneDBName string `env:"SERVER_MYSQL_ONE_DB_NAME"` // 当不为空时，所有mysql资源都将在同一个库中，并且不进行权限控制

	ClientMinioPort   int `env:"CLIENT_MINIO_PORT" env-default:"19000"`
	ClientPostgrePort int `env:"CLIENT_POSTGRE_PORT" env-default:"15432"`
	ClientMySQLPort   int `env:"CLIENT_MYSQL_PORT" env-default:"15433"`
	ClientRedisPort   int `env:"CLIENT_REDIS_PORT" env-default:"16379"`

	LocalTMPPostgrePort int `env:"LOCAL_TMP_POSTGRE_PORT" env-default:"13420"`
	LocalTMPRedisPort   int `env:"LOCAL_TMP_REDIS_PORT" env-default:"13421"`
	LocalTMPMySQLPort   int `env:"LOCAL_TMP_MYSQL_PORT" env-default:"13422"`

	// sso
	SSOAuthURL     string `env:"SSO_AUTH_URL"` // sso认证地址
	SSOCookieName  string `env:"SSO_COOKIE_NAME" env-default:"vv-sso"`
	SSORedirectURL string `env:"SSO_REDIRECT_URL"` // 登录页
	SSOBaseURL     string `env:"SSO_BASE_URL"`     // sso基础地址，如 abc
}

type JwtConfig struct {
	Secret     string `env:"JWT_SECRET" env-default:"secret"`
	ExpireTime int64  `env:"JWT_EXPIRETIME" env-default:"24"` // hour
}

type JwtClaims struct {
	jwt.RegisteredClaims
	UID uint `json:"uid,omitempty"`
}

var (
	AppConfigInstance *AppConfig
	JwtConf           *JwtConfig
	RPCToken          string
)

func init() {
	var err error
	AppConfigInstance = &AppConfig{}
	JwtConf = &JwtConfig{}
	godotenv.Load()

	logrus.Info("env loaded")
	// print all env
	for _, env := range os.Environ() {
		logrus.Info(env)
	}

	if err := cleanenv.ReadEnv(AppConfigInstance); err != nil {
		logrus.Panic(err)
	}
	// print appconfig
	logrus.Info("appconfig loaded")
	logrus.Info(AppConfigInstance)

	if err := cleanenv.ReadEnv(JwtConf); err != nil {
		logrus.Panic(err)
	}

	RPCToken = secret.MD5(fmt.Sprintf("%s%s", AppConfigInstance.NodeName, AppConfigInstance.AgentSecret))
	AppConfigInstance.TunnelUsername = secret.MD5(AppConfigInstance.AgentSecret +
		AppConfigInstance.WorkerURLSuffix)
	AppConfigInstance.TunnelPassword = secret.MD5(AppConfigInstance.AgentSecret +
		AppConfigInstance.WorkerURLSuffix + AppConfigInstance.TunnelUsername)
	AppConfigInstance.TunnelToken = AppConfigInstance.TunnelPassword

	if err != nil {
		logrus.Panic(err)
	}
}

func IsMaster() bool {
	return AppConfigInstance.RunMode == "master"
}
